src/infra/core/src/repos/metadata_block_repository_caching_inmem.rs

```diff
@@ -0,0 +1,94 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use async_trait::async_trait;
+use dashmap::DashMap;
+use kamu_core::{
+    ContainsError,
+    GetBlockError,
+    InsertError,
+    InsertOpts,
+    InsertResult,
+    MetadataBlockRepository,
+    ObjectRepository,
+};
+use opendatafabric::{MetadataBlock, Multihash};
+
+use crate::MetadataBlockRepositoryImpl;
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+pub struct MetadataBlockRepositoryCachingInMem<ObjRepo> {
+    wrapped: MetadataBlockRepositoryImpl<ObjRepo>,
+    cache: DashMap<Multihash, MetadataBlock>,
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+impl<ObjRepo> MetadataBlockRepositoryCachingInMem<ObjRepo>
+where
+    ObjRepo: ObjectRepository + Sync + Send,
+{
+    pub fn new(obj_repo: ObjRepo) -> Self {
+        let wrapped = MetadataBlockRepositoryImpl::new(obj_repo);
+
+        Self {
+            wrapped,
+            cache: DashMap::new(),
+        }
+    }
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait]
+impl<ObjRepo> MetadataBlockRepository for MetadataBlockRepositoryCachingInMem<ObjRepo>
+where
+    ObjRepo: ObjectRepository + Sync + Send,
+{
+    async fn contains(&self, hash: &Multihash) -> Result<bool, ContainsError> {
+        if self.cache.contains_key(hash) {
+            return Ok(true);
+        }
+
+        self.wrapped.contains(hash).await
```

Comment:
Can we implement `contains` via `get`, so that `contains` query caches the block?
In case of a cache miss that still goes to the disk/s3 storage and makes a query of comparable cost as full object retreival (for blocks specificly, as reading 400byes or reading file size is assumed to roughly be the same cost).

Reply:
Absolutely makes sense -- a good catch